base/Base.jl

```diff
@@ -30,7 +30,7 @@ macro noinline() Expr(:meta, :noinline) end
 # (and these don't need the extra convert code)
 getproperty(x::Module, f::Symbol) = (@inline; getglobal(x, f))
 getproperty(x::Type, f::Symbol) = (@inline; getfield(x, f))
-setproperty!(x::Type, f::Symbol, v) = error("setfield! fields of Types should not be changed")
+setproperty!(x::Type, f::Symbol, v) = throw(ArgumentError("setfield! fields of Types should not be changed"))
```

Comment:
Is an `ArgumentError` actually appropriate here? I would read `ArgumentError` as implying that there exists an argument which would not error. The error for this method and the next are warning that this method should not be called ever, regardless of argument, so maybe a `MethodError` would be more appropriate?

Reply:
I don't believe MethodError should ever be thrown outside a dispatch failure according to the type system's rules. If you consider the collective behavior of all methods of the same function, there do exist arguments that work.

Reply:
However, I did definitely stew over whether  `ArgumentError` was right for indicating "this method never works regardless of input". Which is why for now I'm proposing thinking of it from a function view rather than a method view. But open to persuasion there.